docs/ABI/Mangling.rst

```diff
@@ -516,6 +516,10 @@ Types
   FUNCTION-KIND ::= 'C'                      // C function pointer type
   FUNCTION-KIND ::= 'A'                      // @auto_closure function type (escaping)
   FUNCTION-KIND ::= 'E'                      // function type (noescape)
+  FUNCTION-KIND ::= 'F'                      // @differentiable function type
+  FUNCTION-KIND ::= 'G'                      // @differentiable function type (escaping)
+  FUNCTION-KIND ::= 'H'                      // @differentiable(linear) function type
+  FUNCTION-KIND ::= 'I'                      // @differentiable(linear) function type (escaping)
```

Comment:
How would you differentiate between `@differentiable @convention(thin)` and `@differentiable`?

Reply:
Good point. Currently, modifiers like `@convention(thin)` and `@differentiable` have their own function type demangle node variant:

| Modifier | Demangle node kind |
|---|---|
| `-` | `NoEscapeFunctionType` |
| `@escaping` | `FunctionType` (escaping) |
| `@autoclosure` | `AutoClosureType` |
| `@escaping @autoclosure` | `EscapingAutoClosureType` |
| ... | ... |

Here's what it looks like:

```console
$ $SWIFT_BIN/swift-demangle --expand sS2fXfD
Demangling for $sS2fXfD
kind=Global
  kind=TypeMangling
    kind=Type
      kind=ThinFunctionType
        kind=ArgumentTuple
          kind=Type
            kind=Structure
              kind=Module, text="Swift"
              kind=Identifier, text="Float"
        kind=ReturnType
          kind=Type
            kind=Structure
              kind=Module, text="Swift"
              kind=Identifier, text="Float"
$sS2fXfD ---> @convention(thin) (Swift.Float) -> Swift.Float
```

This leads to an explosion of demangle node kinds for combinations of modifiers.

A more reflexible representation would be to have a single `FunctionType` demangle node kind, with modifier nodes as children in a list. Dummy view:

```
      kind=FunctionType
        kind=FunctionTypeModifiers
          kind=ThinFunction
          kind=EscapingFunction
          kind=DifferentiableFunction
        kind=ArgumentTuple
          kind=Type
            kind=Structure
              kind=Module, text="Swift"
              kind=Identifier, text="Float"
        kind=ReturnType
          kind=Type
            kind=Structure
              kind=Module, text="Swift"
              kind=Identifier, text="Float"
```

Changing the mangling scheme certainly seems breaking though. The significance of type mangling isn't exactly clear to me: I imagine LLDB uses this mangling to show types of variables. Perhaps it's not important to capture all combinations of function type modifiers, only the semantically important ones (`@differentiable` is probably more semantically heavy than `@escaping`).

Started a discussion at https://forums.swift.org/t/questions-about-function-type-mangling/26360.